decompiler/pipeline/controlflowanalysis/restructuring_commons/condition_based_refinement.py

```diff
@@ -2,14 +2,164 @@
+from networkx import DiGraph, has_path
+
+
+@dataclass
+class Formula:
+    condition: LogicCondition
+    ast_node: AbstractSyntaxTreeNode
+
+    def __hash__(self) -> int:
+        return id(self)
+
+
+@dataclass
+class Clause:
+    condition: LogicCondition
+    formula: Formula
+
+    def __hash__(self) -> int:
+        return id(self)
+
+
+@dataclass
+class Symbol:
+    name: str
+
+    def __hash__(self):
+        return hash(self.name)
+
+
+class ConditionCandidates:
+    """A graph implementation handling conditions for the condition-based refinement algorithm."""
+
+    def __init__(self, candidates: List[AbstractSyntaxTreeNode]) -> None:
+        self._candidates: Dict[AbstractSyntaxTreeNode, Formula] = {c: Formula(c.reaching_condition, c) for c in candidates}
+        self._logic_graph: DiGraph = DiGraph()
+        self._formulas_containing_symbol: DefaultDict[Symbol, Set[Formula]] = defaultdict(set)
+        self._symbols_of_formula: DefaultDict[Formula, Set[Symbol]] = defaultdict(set)
+        self._initialize_logic_graph_and_dictionaries()
+
+    def _initialize_logic_graph_and_dictionaries(self):
+        for formula in self._candidates.values():
+            self._logic_graph.add_node(formula)
+            formula_clauses = list(formula.condition.operands) if formula.condition.is_conjunction else [formula.condition.copy()]
+            for logic_clause in formula_clauses:
+                self._logic_graph.add_edge(formula, clause := Clause(logic_clause, formula))
+                for symbol_name in logic_clause.get_symbols_as_string():
+                    self._logic_graph.add_edge(clause, symbol := Symbol(symbol_name))
+                    self._formulas_containing_symbol[symbol].add(formula)
+                    self._symbols_of_formula[formula].add(symbol)
+        self._remove_nodes_from(set(symbol for symbol, formulas in self._formulas_containing_symbol.items() if len(formulas) == 1))
+
+    @property
+    def candidates(self) -> Iterator[AbstractSyntaxTreeNode]:
+        yield from self._candidates
+
+    @property
+    def maximum_subexpression_size(self) -> int:
+        if len(self._candidates) < 2:
+            return 0
+        all_sizes = [self._logic_graph.out_degree(formula) for formula in self._candidates.values()]
+        all_sizes.remove(max(all_sizes))
+        return max(all_sizes)
+
+    def get_symbols_of(self, node: AbstractSyntaxTreeNode) -> Set[str]:
+        return {symbol.name for symbol in self._symbols_of_formula[self._candidates[node]]}
+
+    def get_next_subexpression(self) -> Iterator[Tuple[AbstractSyntaxTreeNode, LogicCondition]]:
+        """Consider Candidates in sequence-node order and start with the largest possible subexpression."""
+        for ast_node in list(self._candidates):
+            if ast_node not in self._candidates:
+                continue
+            if (max_expr_size := self.maximum_subexpression_size) == 0:
+                break
```

Comment:
That is exactly the point. Since we group some of the ast-nodes, the maximum expression size can decrease. We could always consider all possible, but if we know it can not be larger than say 2, and we have a condition with 5 literals, we would try 2^5 -1 subexpressions instead of 0.5*5*(5-1) subexpressions.

Reply:
sorry, that was kinda not my point. We are getting the value of a property here and expect that value to change, even though the function checking that contains no functionality which could actually change this value. The iterator is expecting the object it is iteration over to change without any reference to state.

Maybe we should just drop `@property` here and go with get_maximum_subexpression_size()`